README.md
# Computer-Architecture-Assembler-HW3
### Сделал Пересторонин Максим Дмитриевич БПИ 217 Вариант 21 на оценку 6
<br>Из-за требования "необходимо представить только одну
модифицированную ассемблерную программу, разработанную
на целевую оценку. Представление ассемблерных программ, полученных непосредственно после компиляции необходимо только в тех случаях, когда осуществляется их сравнение с модифицированной программой. Разрабатывать и представлять
несколько программ на С и соответствующих им программ на
ассемблере не нужно" многие ссылки введут в одно и то же место. Это связано с тем, что я хотел сохрнить привычную структуру отчета, где видно проделланую работу на каждый пункт того или иного балла.
## На 4 балла:
Для компиляции без модификаций использовалась команда:
<br>`gcc -masm=intel ./<file_name>.c -S -o ./<file_name>.s`
* Приведено [решение задачи на C.](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/CFiles)
* В полученную ассемблерную программу, откомпилированную без оптимизирующих и отладочных опций, добавлены [комментарии](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/AssemblerOriginal), поясняющие эквивалентное представление переменных в программе на C.
* Из ассемблерной программы [убраны лишние макросы](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/AssemblerMod) за счет использования соответствующих аргументов командной строки и/или за счет ручного редактирования исходного текста ассемблерной программы. Модифицированная ассемблерная программа отдельно откомпилирована
и скомпонована без использования опций отладки.
* Представлено [полное тестовое покрытие](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/blob/main/Tests.md), дающее одинаковый результат
на обоих программах. Приведены результаты тестовых прогонов для обоих программ, демонстрирующие эквивалентность функционирования.
* Сформировать отчет, описывающий результаты тестовых прогонов и используемых опций компиляции и/или описания проведенных модификаций. [Заключение по тестам тут](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/blob/main/Tests.md#%D0%B7%D0%B0%D0%BA%D0%BB%D1%8E%D1%87%D0%B5%D0%BD%D0%BD%D0%B8%D0%B5) (смотрите чуть ниже) - [список изменений тут](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/blob/main/AssemblerMod/ChangeLog.md)
## На 5 баллов:
* Я использовал [локальные переменные и передавал данные в функции через параметры](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/CFiles)
* В ассемблерную программу при вызове функции добавить [комментарии, описывающие передачу фактических параметров и перенос возвращаемого результата.](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/AssemblerMod)
* В функциях для формальных параметров добавить [комментарии, описывающие связь между параметрами языка Си и регистрами (стеком)](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/AssemblerMod)
* Отчет: В программе есть функции с передачей данных через параметры, используются локальные переменные. Код прокомментирован.
## На 6 баллов:
* Рефакторинг программы на ассемблере за счет максимального использования регистров процессора. Добавление этой программы к уже представленным. Список изменений [тут](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/blob/main/AssemblerMod/ChangeLog.md), сам код [тут.](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/AssemblerMod)
* Добавление [комментариев в разработанную программу](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/tree/main/AssemblerMod), поясняющих эквивалентное использование регистров вместо переменных исходной программы на C.
* Представление [результатов тестовых прогонов](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/blob/main/Tests.md) для разработанной программы. Оценка корректности ее выполнения на основе сравнения тестовых прогонов результатами тестирования предшествующих программ.
* [Заключение по тестам](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/blob/main/Tests.md#%D0%B7%D0%B0%D0%BA%D0%BB%D1%8E%D1%87%D0%B5%D0%BD%D0%BD%D0%B8%D0%B5)
* Сопоставление [размеров](https://github.com/mperestoronin/Computer-Architecture-Assembler-HW3/blob/main/Size.md) программы на ассемблере, полученной после
компиляции с языка C с модифицированной программой, использующей
регистры.
<br>По всем вопросам писшите на почту mdperestoronin@edu.hse.ru или в телеграмм @Maxim_Perestoronin


AssemblerMod/ChangeLog.md
# Список изменений
* Следующие флаги были применены ко всем файлам:
`gcc -masm=intel -fno-asynchronous-unwind-tables -fno-jump-tables -fno-stack-protector -fno-exceptions -fverbose-asm ./<filename>.c -S -o ./<filename>.s`
* Убран лишний код в main
``` assembly
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"

	.section	.note.GNU-stack,"",@progbits

	.section	.note.gnu.property,"a"

	.align 8

	.long	 1f - 0f

	.long	 4f - 1f

	.long	 5

0:

	.string	 "GNU"

1:

	.align 8

	.long	 0xc0000002

	.long	 3f - 2f

2:

	.long	 0x3

3:

	.align 8

4:
```
* Убран лишний код в getPi
``` assembly
.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4:
```
* Везде поменял параметр i в функции getpi
<br>(QWORD PTR -8[rbp] на xmm3)
Подробнее смотрите непосредственно в ассемблер коде
### Попытка изменить другие регистры привела к некорректной работе программы.


AssemblerMod/main_mod.s
	.file	"main.c"
	.intel_syntax noprefix
	.text
	.section	.rodata
.LC1:
	.string	"Result: %.14f"
	.text
	.globl	main
	.type	main, @function
main:
	endbr64	
	push	rbp	
	mov	rbp, rsp
	sub	rsp, 16
# ./main.c:5:     double pi = 3.0;
	movsd	xmm0, QWORD PTR .LC0[rip]	# переменная pi (double pi = 3)
	movsd	QWORD PTR -8[rbp], xmm0	# переменная pi (double pi = 3)
# ./main.c:6:     printf("Result: %.14f", getPi(pi));
# начало printf("Result: %.14f", getPi(pi));
	mov	rax, QWORD PTR -8[rbp]	# tmp86, pi передача параметра pi
	movq	xmm0, rax	#, tmp86 pi передача параметра pi
	call	getPi@PLT	# вызов getpi
	lea	rdi, .LC1[rip]
	mov	eax, 1	
	call	printf@PLT # вызов printf
	mov	eax, 0	# _6,
# ./main.c:7: }
# конец printf("Result: %.14f", getPi(pi));
	leave	
	ret	
	.size	main, .-main
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1074266112


AssemblerMod/getPi_mod.s
	.file	"getPi.c"
	.intel_syntax noprefix
	.text
	.globl	getPi
	.type	getPi, @function
getPi:
	endbr64	
	push	rbp	
	mov	rbp, rsp
	movsd	QWORD PTR -24[rbp], xmm0	# pi, pi  #берем параметр pi
# ./getPi.c:4:     for (double i = 2.0; i < 3000000; i += 4) {
	movsd	xmm0, QWORD PTR .LC0[rip]	# tmp95,
	movsd	xmm3, xmm0	# i, #tmp95 double i			#меняем QWORD PTR -8[rbp] на xmm3
# ./getPi.c:4:     for (double i = 2.0; i < 3000000; i += 4) {
	jmp	.L2	
.L3:
# ./getPi.c:5:         pi += 4.0 / (i * (i + 1) * (i + 2));
	movsd	xmm1, xmm3	# tmp96, i			#меняем QWORD PTR -8[rbp] на xmm3
	movsd	xmm0, QWORD PTR .LC1[rip]	# tmp97,
	addsd	xmm0, xmm1	# _1, tmp96
# ./getPi.c:5:         pi += 4.0 / (i * (i + 1) * (i + 2));
	movapd	xmm1, xmm0	# _1, _1
	mulsd	xmm1, xmm3	# _1, i			#меняем QWORD PTR -8[rbp] на xmm3
# ./getPi.c:5:         pi += 4.0 / (i * (i + 1) * (i + 2));
	movsd	xmm2, xmm3	# tmp98, i			#меняем QWORD PTR -8[rbp] на xmm3
	movsd	xmm0, QWORD PTR .LC0[rip]	# tmp99,
	addsd	xmm0, xmm2	# _3, tmp98
# ./getPi.c:5:         pi += 4.0 / (i * (i + 1) * (i + 2));
	mulsd	xmm1, xmm0	# _4, _3
# ./getPi.c:5:         pi += 4.0 / (i * (i + 1) * (i + 2));
	movsd	xmm0, QWORD PTR .LC2[rip]	# tmp100,
	divsd	xmm0, xmm1	# _5, _4
# ./getPi.c:5:         pi += 4.0 / (i * (i + 1) * (i + 2));
	movsd	xmm1, QWORD PTR -24[rbp]	# tmp102, pi
	addsd	xmm0, xmm1	# tmp101, tmp102
	movsd	QWORD PTR -24[rbp], xmm0	# pi, tmp101
# ./getPi.c:6:         pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	movsd	xmm1, xmm3	# tmp103, i 			#меняем QWORD PTR -8[rbp] на xmm3
	movsd	xmm0, QWORD PTR .LC0[rip]	# tmp104,
	addsd	xmm1, xmm0	# _6, tmp104
# ./getPi.c:6:         pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	movsd	xmm2, xmm3	# tmp105, i			#меняем QWORD PTR -8[rbp] на xmm3
	movsd	xmm0, QWORD PTR .LC3[rip]	# tmp106,
	addsd	xmm0, xmm2	# _7, tmp105
# ./getPi.c:6:         pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	mulsd	xmm1, xmm0	# _8, _7
# ./getPi.c:6:         pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	movsd	xmm2, xmm3	# tmp107, i			#меняем QWORD PTR -8[rbp] на xmm3
	movsd	xmm0, QWORD PTR .LC2[rip]	# tmp108,
	addsd	xmm0, xmm2	# _9, tmp107
# ./getPi.c:6:         pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	mulsd	xmm1, xmm0	# _10, _9
# ./getPi.c:6:         pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	movsd	xmm0, QWORD PTR .LC2[rip]	# tmp109,
	divsd	xmm0, xmm1	# tmp109, _10
	movapd	xmm1, xmm0	# _11, tmp109
# ./getPi.c:6:         pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	movsd	xmm0, QWORD PTR -24[rbp]	# tmp111, pi
	subsd	xmm0, xmm1	# tmp110, _11
	movsd	QWORD PTR -24[rbp], xmm0	# pi, tmp110
# ./getPi.c:4:     for (double i = 2.0; i < 3000000; i += 4) {
	movsd	xmm1, xmm3	# tmp113, i			#меняем QWORD PTR -8[rbp] на xmm3
	movsd	xmm0, QWORD PTR .LC2[rip]	# tmp114,
	addsd	xmm0, xmm1	# tmp112, tmp113
	movsd	xmm3, xmm0	# i, tmp112			#меняем QWORD PTR -8[rbp] на xmm3
.L2:
# ./getPi.c:4:     for (double i = 2.0; i < 3000000; i += 4) {
	movsd	xmm0, QWORD PTR .LC4[rip]	# tmp115,
	comisd	xmm0, xmm3	# tmp115, i			#меняем QWORD PTR -8[rbp] на xmm3
	ja	.L3	
# ./getPi.c:8:     return pi;
	movsd	xmm0, QWORD PTR -24[rbp]	# _16, pi # возвращаем pi
# ./getPi.c:9: }
	pop	rbp	#
	ret	
	.size	getPi, .-getPi
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1073741824
	.align 8
.LC1:
	.long	0
	.long	1072693248
	.align 8
.LC2:
	.long	0
	.long	1074790400
	.align 8
.LC3:
	.long	0
	.long	1074266112
	.align 8
.LC4:
	.long	0
	.long	1095164768


AssemblerOriginal/main.s
	.file	"main.c"
	.intel_syntax noprefix
	.text
	.section	.rodata
.LC1:
	.string	"Result: %.14f" # строка из строчки printf("Result: %.14f", getPi(pi));
	.text
	.globl	main
	.type	main, @function
main:
.LFB0:
	.cfi_startproc
	endbr64
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	mov	rbp, rsp
	.cfi_def_cfa_register 6
	sub	rsp, 16
	movsd	xmm0, QWORD PTR .LC0[rip] # переменная pi (double pi = 3.0)
	movsd	QWORD PTR -8[rbp], xmm0 # переменная pi (double pi = 3.0)
	# передача параметров и вызов getpi
	mov	rax, QWORD PTR -8[rbp] #передаем pi  в функцию getPi(pi)
	movq	xmm0, rax
	call	getPi@PLT # вызов getPi
	lea	rdi, .LC1[rip]
	mov	eax, 1
	call	printf@PLT # вызов printf
	mov	eax, 0 # return 0 в рузультате успешной работы программы
	leave
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE0:
	.size	main, .-main
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1074266112
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4:


AssemblerOriginal/getPi.s
	.file	"getPi.c"
	.intel_syntax noprefix
	.text
	.globl	getPi
	.type	getPi, @function
getPi:
.LFB0:
	.cfi_startproc
	endbr64
	push	rbp
	.cfi_def_cfa_offset 16
	.cfi_offset 6, -16
	mov	rbp, rsp
	.cfi_def_cfa_register 6
	movsd	QWORD PTR -24[rbp], xmm0 # double pi
	movsd	xmm0, QWORD PTR .LC0[rip]
	movsd	QWORD PTR -8[rbp], xmm0 # double i
	jmp	.L2 # цикл for
.L3:# начало pi += 4.0 / (i * (i + 1) * (i + 2));
	movsd	xmm1, QWORD PTR -8[rbp]
	movsd	xmm0, QWORD PTR .LC1[rip]
	addsd	xmm0, xmm1
	movapd	xmm1, xmm0
	mulsd	xmm1, QWORD PTR -8[rbp] # кстати mulsd значит Multiply Scalar Double-Precision Floating-Point Value, т.е. перемножение чисел с плавающей точкой
	movsd	xmm2, QWORD PTR -8[rbp]
	movsd	xmm0, QWORD PTR .LC0[rip]
	addsd	xmm0, xmm2
	mulsd	xmm1, xmm0
	movsd	xmm0, QWORD PTR .LC2[rip]
	divsd	xmm0, xmm1 # деление 4 / (то, что считалось ранее)
	movsd	xmm1, QWORD PTR -24[rbp] # присваиваем старое значение pi
	addsd	xmm0, xmm1 # складываем старое значение pi с результатами вычилений 
	movsd	QWORD PTR -24[rbp], xmm0 # присваивание результата операции обратно в переменную pi
	# конец pi += 4.0 / (i * (i + 1) * (i + 2));
	# начало pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	movsd	xmm1, QWORD PTR -8[rbp]
	movsd	xmm0, QWORD PTR .LC0[rip]
	addsd	xmm1, xmm0
	movsd	xmm2, QWORD PTR -8[rbp]
	movsd	xmm0, QWORD PTR .LC3[rip]
	addsd	xmm0, xmm2
	mulsd	xmm1, xmm0
	movsd	xmm2, QWORD PTR -8[rbp]
	movsd	xmm0, QWORD PTR .LC2[rip]
	addsd	xmm0, xmm2
	mulsd	xmm1, xmm0
	movsd	xmm0, QWORD PTR .LC2[rip]
	divsd	xmm0, xmm1
	movapd	xmm1, xmm0
	movsd	xmm0, QWORD PTR -24[rbp]
	subsd	xmm0, xmm1
	movsd	QWORD PTR -24[rbp], xmm0
	# конец pi -= 4.0 / ((i + 2) * (i + 3) * (i + 4));
	movsd	xmm1, QWORD PTR -8[rbp]
	movsd	xmm0, QWORD PTR .LC2[rip]
	addsd	xmm0, xmm1
	movsd	QWORD PTR -8[rbp], xmm0
.L2:
	movsd	xmm0, QWORD PTR .LC4[rip]
	comisd	xmm0, QWORD PTR -8[rbp]
	ja	.L3
	movsd	xmm0, QWORD PTR -24[rbp] # return pi
	pop	rbp
	.cfi_def_cfa 7, 8
	ret
	.cfi_endproc
.LFE0:
	.size	getPi, .-getPi
	.section	.rodata
	.align 8
.LC0:
	.long	0
	.long	1073741824
	.align 8
.LC1:
	.long	0
	.long	1072693248
	.align 8
.LC2:
	.long	0
	.long	1074790400
	.align 8
.LC3:
	.long	0
	.long	1074266112
	.align 8
.LC4:
	.long	0
	.long	1095164768
	.ident	"GCC: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0"
	.section	.note.GNU-stack,"",@progbits
	.section	.note.gnu.property,"a"
	.align 8
	.long	 1f - 0f
	.long	 4f - 1f
	.long	 5
0:
	.string	 "GNU"
1:
	.align 8
	.long	 0xc0000002
	.long	 3f - 2f
2:
	.long	 0x3
3:
	.align 8
4:
